In [3]:
# !pip install ydata-profiling

In [4]:
import numpy as np
import pandas as pd
import seaborn as sns
import os
import time
import requests
import tarfile
import matplotlib.pyplot as plt
%matplotlib inline
# from ydata_profiling import ProfileReport

In [6]:
import os

current_dir = os.getcwd()

In [7]:
df_top10 = pd.read_csv('./DATA/netflix daily top 10.csv')

df_titles = pd.read_csv('./DATA/netflix_titles.csv')

In [8]:
df_top10.head()


As of  Rank Year to Date Rank Last Week Rank  \
0  2020-04-01     1                 1              1   
1  2020-04-01     2                 2              -   
2  2020-04-01     3                 3              2   
3  2020-04-01     4                 4              -   
4  2020-04-01     5                 5              4   

                          Title     Type Netflix Exclusive  \
0  Tiger King: Murder, Mayhem …  TV Show               Yes   
1                         Ozark  TV Show               Yes   
2                  All American  TV Show               NaN   
3                  Blood Father    Movie               NaN   
4                  The Platform    Movie               Yes   

  Netflix Release Date  Days In Top 10  Viewership Score  
0         Mar 20, 2020               9                90  
1         Jul 21, 2017               5                45  
2         Mar 28, 2019               9                76  
3         Mar 26, 2020               5                30  
4         Mar 20, 2020               9                55

In [9]:
df_titles.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

Skipping the use of pandas Profiling for the purposes of this final report notebook. This notebook is to generate the processed data to use in the final models. In development work, this notebook also contained exploratory data analysis with pandas profiling.

In [10]:
# profile_top10 = ProfileReport(df_top10, title='Netflix Top 10 April 2020 to March 2022')

In [11]:
# profile_top10.to_notebook_iframe()

In [12]:
# profile_titles = ProfileReport(df_titles, title='Netflix Titles')
# profile_titles.to_notebook_iframe()

# Exploring More Data

Lets explore some other data... starting with two IMDB datasets from here: https://datasets.imdbws.com/title.ratings.tsv.gz','https://datasets.imdbws.com/title.basics.tsv.gz%20using%20imdb%20data
For the purposes of the final report, the team opted to not utilize this data, so it is commented out of this notebook. For dev notebook, please see: https://github.com/Examining-Netflix-Top-Ten-Feature/final-project/blob/main/datapreprocessing_dev.ipynb

In [13]:
# urls = ['https://datasets.imdbws.com/title.ratings.tsv.gz','https://datasets.imdbws.com/title.basics.tsv.gz']
# for url in urls:
#     r=requests.get(url)
#     with open(url.split('/')[-1], 'wb') as fd:
#         for chunk in r.iter_content(chunk_size=128):
#             fd.write(chunk)

In [14]:
# basics_df = pd.read_csv('title.basics.tsv.gz',sep='\t')
# ratings_df = pd.read_csv('title.ratings.tsv.gz',sep='\t')

# Join top 10 dataset with the netflix titles dataset 
(instead of the IMDB data)



In [20]:
df_full = pd.merge(df_top10, df_titles, left_on= "Title", right_on="title", how='left')

In [21]:
df_full.head(20)

As of  Rank Year to Date Rank Last Week Rank  \
0   2020-04-01     1                 1              1   
1   2020-04-01     2                 2              -   
2   2020-04-01     3                 3              2   
3   2020-04-01     4                 4              -   
4   2020-04-01     5                 5              4   
5   2020-04-01     6                 6              -   
6   2020-04-01     7                10              -   
7   2020-04-01     8                 7              5   
8   2020-04-01     9                 8              -   
9   2020-04-01    10                 9              -   
10  2020-04-02     1                 1              1   
11  2020-04-02     2                 2              -   
12  2020-04-02     3                 3              2   
13  2020-04-02     4                 -              -   
14  2020-04-02     5                 -              -   
15  2020-04-02     6                 4              -   
16  2020-04-02     7                 -              -   
17  2020-04-02     8                 7              -   
18  2020-04-02     9                 5              4   
19  2020-04-02    10                 6              -   

                           Title     Type Netflix Exclusive  \
0   Tiger King: Murder, Mayhem …  TV Show               Yes   
1                          Ozark  TV Show               Yes   
2                   All American  TV Show               NaN   
3                   Blood Father    Movie               NaN   
4                   The Platform    Movie               Yes   
5    Car Masters: Rust to Riches  TV Show               Yes   
6                     Unorthodox  TV Show               Yes   
7                  Love is Blind  TV Show               Yes   
8                        Badland    Movie               NaN   
9                       Uncorked    Movie               Yes   
10  Tiger King: Murder, Mayhem …  TV Show               Yes   
11                         Ozark  TV Show               Yes   
12                  All American  TV Show               NaN   
13                    Nailed It!  TV Show               Yes   
14     How to Fix a Drug Scandal  TV Show               Yes   
15                  Blood Father    Movie               NaN   
16                  The Roommate    Movie               NaN   
17                    Unorthodox  TV Show               Yes   
18                  The Platform    Movie               Yes   
19   Car Masters: Rust to Riches  TV Show               Yes   

   Netflix Release Date  Days In Top 10  Viewership Score  ...  \
0          Mar 20, 2020               9                90  ...   
1          Jul 21, 2017               5                45  ...   
2          Mar 28, 2019               9                76  ...   
3          Mar 26, 2020               5                30  ...   
4          Mar 20, 2020               9                55  ...   
5          Sep 14, 2018               4                14  ...   
6          Mar 26, 2020               2                 5  ...   
7          Feb 13, 2020               9                40  ...   
8          Mar 26, 2020               4                11  ...   
9          Mar 27, 2020               4                15  ...   
10         Mar 20, 2020              10               100  ...   
11         Jul 21, 2017               6                54  ...   
12         Mar 28, 2019              10                84  ...   
13          Mar 9, 2018               1                 7  ...   
14          Apr 1, 2020               1                 6  ...   
15         Mar 26, 2020               6                35  ...   
16          Apr 1, 2020               1                 4  ...   
17         Mar 26, 2020               3                 8  ...   
18         Mar 20, 2020              10                57  ...   
19         Sep 14, 2018               5                15  ...   

                          title                   director  \
0                           NaN       

In [22]:
df_full['title'].isnull().values.any

<function ndarray.any>

In [23]:
df_full['title'].isnull().sum()

2144

In [24]:
df_full['Title'].isnull().sum()

0

In [25]:
df_full['Title'].nunique()


645

In [26]:
df_full['title'].nunique()

412

There are 645 unique titles in the top 10 dataset. 412 titles were found in the joined titles data set. so about 200 are missing the additional titles,genre,director, etc.

In [27]:
print(412/645*100)

63.87596899224807


The titles feature from the top ten data set has no nulls. but after joining with the titles data set, the titles field has 2144 null values. In other words, there are 2144 titles that are in the top 10 data set but are not found in the titles data set.


In [28]:
df_full.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7100 entries, 0 to 7099
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   As of                 7100 non-null   object 
 1   Rank                  7100 non-null   int64  
 2   Year to Date Rank     7100 non-null   object 
 3   Last Week Rank        7100 non-null   object 
 4   Title                 7100 non-null   object 
 5   Type                  7100 non-null   object 
 6   Netflix Exclusive     4599 non-null   object 
 7   Netflix Release Date  7100 non-null   object 
 8   Days In Top 10        7100 non-null   int64  
 9   Viewership Score      7100 non-null   int64  
 10  show_id               4956 non-null   object 
 11  type                  4956 non-null   object 
 12  title                 4956 non-null   object 
 13  director              1885 non-null   object 
 14  cast                  4211 non-null   object 
 15  country              

In [29]:
df_full['as_of_date'] = pd.to_datetime(df_full['As of'])
df_full['netflix_release_date'] = pd.to_datetime(df_full['Netflix Release Date'])
df_full['Netflix Exclusive'] = df_full['Netflix Exclusive'].fillna('NO')
df_full['release_year'] = df_full['release_year'].astype('Int64')


In [30]:
df_full.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7100 entries, 0 to 7099
Data columns (total 24 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   As of                 7100 non-null   object        
 1   Rank                  7100 non-null   int64         
 2   Year to Date Rank     7100 non-null   object        
 3   Last Week Rank        7100 non-null   object        
 4   Title                 7100 non-null   object        
 5   Type                  7100 non-null   object        
 6   Netflix Exclusive     7100 non-null   object        
 7   Netflix Release Date  7100 non-null   object        
 8   Days In Top 10        7100 non-null   int64         
 9   Viewership Score      7100 non-null   int64         
 10  show_id               4956 non-null   object        
 11  type                  4956 non-null   object        
 12  title                 4956 non-null   object        
 13  director          

In [31]:
#feature engineer for the first full dataframe (which uses the netflix titles csv -- ie not IMDB)
df_full['age_days_since_rel'] = df_full['as_of_date'] - df_full['netflix_release_date']

In [32]:
df_full.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7100 entries, 0 to 7099
Data columns (total 25 columns):
 #   Column                Non-Null Count  Dtype          
---  ------                --------------  -----          
 0   As of                 7100 non-null   object         
 1   Rank                  7100 non-null   int64          
 2   Year to Date Rank     7100 non-null   object         
 3   Last Week Rank        7100 non-null   object         
 4   Title                 7100 non-null   object         
 5   Type                  7100 non-null   object         
 6   Netflix Exclusive     7100 non-null   object         
 7   Netflix Release Date  7100 non-null   object         
 8   Days In Top 10        7100 non-null   int64          
 9   Viewership Score      7100 non-null   int64          
 10  show_id               4956 non-null   object         
 11  type                  4956 non-null   object         
 12  title                 4956 non-null   object         
 13  dir

In [33]:
df_full.shape

(7100, 25)

# Export to CSV


In [61]:
#export data to csv for this first full dataframe (not the imdb dataframe)
df_full.to_csv('./DATA/netflix_and_titles_ds1.csv', sep=',', encoding='utf-8')

In [35]:
df_full['country'].fillna(0, inplace=True)
df_full['listed_in'].fillna(0, inplace=True)
df_full.head()


/var/folders/6x/51znw2017v35lmzlybgk2fgc0000gp/T/ipykernel_2027/159771968.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_full['country'].fillna(0, inplace=True)
/var/folders/6x/51znw2017v35lmzlybgk2fgc0000gp/T/ipykernel_2027/159771968.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves

As of  Rank Year to Date Rank Last Week Rank  \
0  2020-04-01     1                 1              1   
1  2020-04-01     2                 2              -   
2  2020-04-01     3                 3              2   
3  2020-04-01     4                 4              -   
4  2020-04-01     5                 5              4   

                          Title     Type Netflix Exclusive  \
0  Tiger King: Murder, Mayhem …  TV Show               Yes   
1                         Ozark  TV Show               Yes   
2                  All American  TV Show                NO   
3                  Blood Father    Movie                NO   
4                  The Platform    Movie               Yes   

  Netflix Release Date  Days In Top 10  Viewership Score  ...  \
0         Mar 20, 2020               9                90  ...   
1         Jul 21, 2017               5                45  ...   
2         Mar 28, 2019               9                76  ...   
3         Mar 26, 2020               5                30  ...   
4         Mar 20, 2020               9                55  ...   

                country      date_added release_year rating   duration  \
0                     0             NaN         <NA>    NaN        NaN   
1         United States  March 27, 2020         2020  TV-MA  3 Seasons   
2         United States   July 27, 2021         2021  TV-14  3 Seasons   
3                France  March 26, 2020         2016      R     88 min   
4  United Arab Emirates    June 2, 2021         2021  TV-14  3 Seasons   

                                         listed_in  \
0                                                0   
1          Crime TV Shows, TV Dramas, TV Thrillers   
2                         TV Dramas, Teen TV Shows   
3         Action & Adventure, International Movies   
4  International TV Shows, TV Dramas, TV Mysteries   

                                         description  as_of_date  \
0                                                NaN  2020-04-01   
1  A financial adviser drags his family from Chic...  2020-04-01   
2  Culture clashes and brewing rivalries test a t...  2020-04-01   
3  An ex-convict and estranged father finds the c...  2020-04-01   
4  A programming genius builds a fact-finding, tr...  2020-04-01   

  netflix_release_date age_days_since_rel  
0           2020-03-20            12 days  
1           2017-07-21           985 days  
2           2019-03-28           370 days  
3           2020-03-26             6 days  
4           2020-03-20            12 days  

[5 rows x 25 columns]

In [36]:
# Splitting the genres into separate columns at commas
genres_split = df_full['listed_in'].str.split(', ', expand=True)

# Creating binary columns for each unique value
genres_dummies = pd.get_dummies(genres_split.stack()).groupby(level=0).sum()

# Concatenating the binary columns with the original DataFrame
df_full = pd.concat([df_full, genres_dummies], axis=1)

# Drop Listed in
# df_full.drop('listed_in', axis=1, inplace=True)

#
# df_full

In [37]:
# Extract numerical values and units into separate columns
df_full['duration_minutes'] = df_full['duration'].str.extract(r'(\d+) min', expand=False).astype(float)
df_full['duration_seasons'] = df_full['duration'].str.extract(r'(\d+) Season[s]?', expand=False).astype(float)

# Fill NaN values in case there are no seasons or minutes mentioned
df_full['duration_minutes'].fillna(0, inplace=True)
df_full['duration_seasons'].fillna(0, inplace=True)


/var/folders/6x/51znw2017v35lmzlybgk2fgc0000gp/T/ipykernel_2027/1084383383.py:6: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_full['duration_minutes'].fillna(0, inplace=True)
/var/folders/6x/51znw2017v35lmzlybgk2fgc0000gp/T/ipykernel_2027/1084383383.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values alw

In [54]:
# Splitting the country into separate columns at commas
country_split = df_full['country'].str.split(', ', expand=True)

# Creating binary columns for each unique value
country_dummies = pd.get_dummies(country_split.stack()).groupby(level=0).sum()

# Concatenating the binary columns with the original DataFrame
df_full = pd.concat([df_full, country_dummies], axis=1)


# Generate a new dataset to see the country of production for each program. This is a flattened dataset.

In [ ]:
df_ext = df_full[['Title', 'As of', 'Australia', 'Belgium', 'Brazil', 'Bulgaria', 'Cambodia', 'Canada', 'Chile', 'China', 'Colombia', 'Denmark', 'France', 'Germany', 'Greece', 'Hong Kong', 'Hungary', 'India', 'Israel', 'Italy', 'Japan', 'Luxembourg', 'Malta', 'Mauritius', 'Mexico', 'Norway', 'Poland', 'Romania', 'Russia', 'Singapore', 'South Africa', 'South Korea', 'Spain', 'Sweden', 'Thailand', 'United Arab Emirates', 'United Kingdom', 'United States']]
df_ext.head()

In [57]:
#melt the df to flatten the country field...
# df_ext.head(20)
df_melted = df_ext.melt( id_vars= ["Title", "As of"], var_name = "Country", value_name = "Country_Count")
df_melted['Country_Count'].fillna(0, inplace=True)
df_melted.head()

/var/folders/6x/51znw2017v35lmzlybgk2fgc0000gp/T/ipykernel_2027/3884982020.py:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_melted['Country_Count'].fillna(0, inplace=True)


Title       As of    Country  Country_Count
0  Tiger King: Murder, Mayhem …  2020-04-01  Australia            0.0
1                         Ozark  2020-04-01  Australia            0.0
2                  All American  2020-04-01  Australia            0.0
3                  Blood Father  2020-04-01  Australia            0.0
4                  The Platform  2020-04-01  Australia            0.0

# Export to a csv

In [60]:
df_melted.to_csv('./DATA/netflix_and_titles_melted_countries1.csv', sep=',', encoding='utf-8')

In [62]:
#melt the df to flatten the genre field...
df_g_ext = df_full[["Title", "As of",  "Action & Adventure", "Anime Series", "British TV Shows", "Children & Family Movies", "Classic & Cult TV", "Classic Movies", "Comedies", "Crime TV Shows", "Documentaries", "Docuseries", "Dramas", "Faith & Spirituality", "Independent Movies", "International Movies", "International TV Shows", "Kids' TV", "Korean TV Shows", "LGBTQ Movies", "Music & Musicals", "Reality TV", "Romantic Movies", "Romantic TV Shows", "Sci-Fi & Fantasy", "Science & Nature TV", "Spanish-Language TV Shows", "Sports Movies", "Stand-Up Comedy & Talk Shows", "Stand-Up Comedy", "Teen TV Shows", "Thrillers", "TV Action & Adventure", "TV Comedies", "TV Dramas", "TV Horror", "TV Mysteries", "TV Sci-Fi & Fantasy", "TV Thrillers"]]
# df_g_ext.head(20)
df_melted_g = df_g_ext.melt( id_vars= ["Title", "As of"], var_name = "Genre", value_name = "Genre_Count")
df_melted_g['Genre_Count'].fillna(0, inplace=True)
df_melted_g.head()

/var/folders/6x/51znw2017v35lmzlybgk2fgc0000gp/T/ipykernel_2027/92409771.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_melted_g['Genre_Count'].fillna(0, inplace=True)


Title       As of               Genre  Genre_Count
0  Tiger King: Murder, Mayhem …  2020-04-01  Action & Adventure          0.0
1                         Ozark  2020-04-01  Action & Adventure          0.0
2                  All American  2020-04-01  Action & Adventure          0.0
3                  Blood Father  2020-04-01  Action & Adventure          1.0
4                  The Platform  2020-04-01  Action & Adventure          0.0

# Export to a csv

In [63]:
df_melted_g.to_csv('./DATA/netflix_and_titles_melted_genres1.csv', sep=',', encoding='utf-8')

END
